2단계 - 문자열 계산기

hiddenparts opened the pull request:
급하게 구현하다 보니 놓친 부분이 많을것 같습니다.
기능 구현에 집중하다보니 구현을 완성하고 그 뒤에 테스트케이스를 작성하여 TDD에 맞게 코딩을 했는지 의문점이 생깁니다.
그래도 테스트케이스를 뒤늦게 작성하였지만 구현할때는 놓쳤던 구멍들을 찾아낼 수 있어서 테스트코드를 왜 작성해야하는지 알 수 있는 좋은 시간이었습니다.

hiddenparts committed:
사용하지않아 불필요한 함수 삭제

hiddenparts committed:
함수명을 좀 더 명확하게 하고 테스트케이스 구현방식 수정

hiddenparts committed:
셋에 대한 학습 테스트

hiddenparts committed:
[2단계] 사칙연산 수행 기능 구현

- 더하기, 빼기, 곱하기, 나누기 기능 구현

hiddenparts committed:
[2단계] 입력문자열을 분리하는 기능 구현

hiddenparts committed:
[2단계] 요구사항 구현

- 입력값 구분
- 각 연산자에 따른 함수 호출
- 마지막에 결과 출력
- 함수를 주석으로 목적별로 분리

hiddenparts committed:
[2단계] 예외처리

- 에러 정의
- 에러가 발생될 수 있는 부분 설정

hiddenparts committed:
[2단계] 리턴부분 정리

- 결과를 외부에서 받아볼 수 있도록 함

hiddenparts committed:
[2단계] 실제 계산함수 분리

hiddenparts committed:
[2단계] 연산자 부분 분리

- 에러코드 세분화

hiddenparts committed:
[2단계] 입력값 처리함수 분리

hiddenparts committed:
[2단계] 리팩토링

- if만 사용, else 제거
- 1단계 들여쓰기만 사용
- 좀 더 명확한 이름으로 함수명 변경

hiddenparts committed:
[2단계] 테스트케이스 작성

- 테스트를 위해 readline 함수 제거
- readline이 없어지면서 리팩토링
- 기존의 테스트 무시하도록 Target 제거

hiddenparts committed:
[2단계] 테스트케이스 작성

- 잘못된 입력값에 대한 테이스케이스 작성

hiddenparts committed:
[2단계] 사칙연산 테스트케이스 작성

uuu1101 commented on RacingCar/Calculator.swift:
```diff
@@ -0,0 +1,100 @@
+import Foundation
+
+enum CalculatorError: Error {
+    case noInput
+    case notNumber  // 숫자가 없다
+    case notOperator   // 연산자가 없다
+    case numberNotExist
+    case operatorNotExist
+    case inputIsNotValid
+    case divideByZero
```
주석의 위치가 올바른지 확인부탁드립니다 :) 
주석이 코드를 이해하는데 도움을 주지만, 현재는 오해를 불러 일으킬 수도 있을 것 같습니다 
enum 으로 Error 타입을 구현하셨네요. 👍👍👍

hiddenparts replied:
너무 급하게 한 티가 팍팍 나네요😓
쓰지말던가 주석을 다 달던가를 했었어야 했는데 이부분은 수정하겠습니다!

uuu1101 commented on RacingCar/Calculator.swift:
```diff
@@ -0,0 +1,100 @@
+struct Calculator {
+    
+    var result: Int?
+    var opCode: String?
```
Swift의 [API Design Guideline](https://swift.org/documentation/api-design-guidelines/#naming) 문서를 보았을 때  ~~`opCode` 보다는 `operationCode` 라고 풀어쓰는게 문서에서 요구하는것에 가깝지 않을까? 라는 생각을 잠시 해보았는데요.~~ 한번 고민해보는건 어떨까 해서 피드백드립니다.
어떻게 생각하시나요? 🤔

uuu1101 replied:
전체 코드를 보니, opCode를 보니 연산자를 나타내는 것 같네요.
명령코드를 나타내는 opCode(operationCode) 보다는 연산자 < 라는 의미를 가진 단어로 바꾸어 주는게 좋아보이는데 어떻게 생각하시나요??👀

hiddenparts replied:
연산자를 저장하는 변수로 처음에 opeartor로 썼다가 이게 예약어여서 비슷한 의미를 가진 opCode로 사용을 했네요. op라고 하면 너무 줄이는거 같아서 Code를 붙였는데 더 좋은 변수명을 고민해보도록 하겠습니다

uuu1101 commented on RacingCar/Calculator.swift:
```diff
@@ -0,0 +1,100 @@
+    
+    /// 동작
+    mutating func input(str: String) throws -> Int {
+        let inputs = str.trimmingCharacters(in: .whitespacesAndNewlines).components(separatedBy: " ")
```
능숙하게 문자열을 다루셨네요 👍  , split 과 componets의 동작이 유사한데 혹시 components를 선택하신 이유가 있을까요?

hiddenparts replied:
강의에서 hint로 주어진게 components여서 별 생각없이 사용하였네요 😅

uuu1101 commented on RacingCar/Calculator.swift:
```diff
@@ -0,0 +1,100 @@
+    private mutating func drive(inputs: [String]) throws -> Int {
+        guard inputs.count > 2 else { throw CalculatorError.inputIsNotValid }
+        
+        var inputs = inputs
```
혹시 이 코드는 어떤의미가 담겨있는지 알려주실수 있나요?🤔

hiddenparts replied:
입력을 통해 쪼개진 배열이 식을 이루기위해선 개수가 3개 이상이어야 하는데 그보다 적을 경우 에러를 리턴하도록 했습니다.
근데 지금 다시 보니 
`count > 2` 보다는 `count >= 3` 이 좀더 직관적일거 같은 느낌이 드네요🤔

uuu1101 commented on RacingCar/Calculator.swift:
```diff
@@ -0,0 +1,100 @@
+        return try drive(inputs: inputs)
+    }
+    
+    private mutating func drive(inputs: [String]) throws -> Int {
```
전체 코드 흐름이 시작되는 부분으로 보여지는데, drive 라고 이름을 지으신 이유가 있나요?

hiddenparts replied:
별 뜻은 없고 메인 로직을 실행하는 부분을 start나 drive 이런 표현으로 즐겨써서 그렇습니다.

uuu1101 commented on RacingCar/Calculator.swift:
```diff
@@ -0,0 +1,100 @@
+        opCode = parseOpcode(input: input)
+    }
+    
+    private mutating func parseNumber(input: String) -> Int? {
```
함수에서 Int(input)만을 하고 있는데, 이렇게 함수로 분리하신 이유가 있을까요?

uuu1101 commented on RacingCar/Calculator.swift:
```diff
@@ -0,0 +1,100 @@
+        return result ?? 0
+    }
+    
+    private mutating func processInput(input: String) throws {
```
함수명만으로 이 함수가 어떤일을 하는지 명확하게 알 수 있으면 좋을것 같아요

uuu1101 commented on RacingCar/Calculator.swift:
```diff
@@ -0,0 +1,100 @@
+        return Int(input)
+    }
+    
+    private func parseOpcode(input: String) -> String? {
```
연산자를 판단하고 반환하는 함수로 보여지네요 , 파싱보다는 판단하는 의미를 가지는게 더 좋지 않을까 라는 생각이 듭니다.

uuu1101 commented on RacingCar/Calculator.swift:
```diff
@@ -0,0 +1,100 @@
+        return nil
+    }
+    
+    private mutating func calc(previous: Int, opCode: String, number: Int) throws -> Int? {
```
여기도 축약어가 보이네요, [API Design Guideline](https://swift.org/documentation/api-design-guidelines/#naming)를 참고하시면 좋을듯 합니다 😀

uuu1101 commented on RacingCar/Calculator.swift:
```diff
@@ -0,0 +1,100 @@
+    }
+    
+    /// 기능
+    func add(_ lhs: Int, _ rhs: Int) -> Int {
+        return lhs + rhs
+    }
+    
+    func substract(_ lhs: Int, _ rhs: Int) -> Int {
+        return lhs - rhs
+    }
+    
+    func multiply(_ lhs: Int, _ rhs: Int) -> Int {
+        return lhs * rhs
+    }
+    
+    func divide(_ lhs: Int, _ rhs: Int) throws -> Int {
+        if rhs == 0 {
+            throw CalculatorError.divideByZero
+        }
+        return lhs / rhs
+    }
```
함수명이 같은 컨벤션을 가지고 네이밍 된 것 같아서 좋네요 ! 👍 의미도 명확하구요 !

hiddenparts replied:
감사합니다 😄

uuu1101 commented on RacingCar/main.swift:
```diff
@@ -5,4 +5,3 @@
 // 
 
 import Foundation
```
이 파일은 왜 생성되어 있을까요?

hiddenparts replied:
main.swift 파일은 삭제해도 문제가 없는걸까요? macOS의 commandline project의 엔트리포인트라 삭제하면 안되는 걸로 알고있습니다.

uuu1101 commented on RacingCarTest/CalculatorTest.swift:
```diff
@@ -0,0 +1,35 @@
+class CalculatorTest: XCTestCase {
+
+    func testValidInput() throws {
+        var cal = Calculator()
```
테스트 코드이지만 축약형보다는 풀어쓰는게 좋아보입니다 !

hiddenparts replied:
함수명을 너무 축약해서 적었단 뜻일까요?

uuu1101 replied:
Calculator를 나타내는 cal에 대한 이야기 였습니다 😄

uuu1101 commented on RacingCarTest/CalculatorTest.swift:
```diff
@@ -0,0 +1,35 @@
+        let input = "2 + 3 * 4 / 2"
+        let result = try cal.input(str: input)
+        
+        XCTAssertEqual(result, 10)
```
같지 않은 결과를 가진 테스트는 어떤 메서드를 사용하면 좋을까요?

hiddenparts replied:
XCTAssertNotEqual를 사용하여 더 많은 테스트를 해봤음 좋을텐데 제가 너무 테스트를 대충 짰네요;

uuu1101 commented on RacingCarTest/CalculatorTest.swift:
```diff
@@ -0,0 +1,35 @@
+        XCTAssertEqual(result, 10)
+    }
+    
+    func testInvalidInput() throws {
+        var cal = Calculator()
+        let input = "1 + "
+        XCTAssertThrowsError(try cal.input(str: input))
+    }
+    
+    func test_사칙연산() throws {
```
위는 메서드명을 카멜케이스로 영어로 작성을 하셨는데, 아래는 한글로 스네이크 케이스로 메서드명을 작성하셨네요.
이렇게 작성하신 이유가 궁금합니다 ! 😄

hiddenparts replied:
여..영작이 힘들어서(...)

uuu1101 replied:
아,그런 의미는 아니였어요 ㅎㅎ 
두 메서드명의 컨벤션이 달라서 여쭤본거에요 😊

uuu1101 reviewed (changes requested):
안녕하세요. 반갑습니다~

첫 PR을 보내시느라 고생하셨습니다.👏

전반적으로 네이밍에 대해서 고민을 해보셨으면 좋겠습니다.

피드백 남겨두었으니, 확인하시고
궁금하신 부분 있다면 댓글 또는 DM 보내주세요.😀

hiddenparts committed:
[2단계] 피드백 반영

- 어떠한 에러인지 주석 작성
- drive 함수명을 좀 더 명확하게 변경
- opCode의 이름을 좀 더 명확하게 변경
- 테스트 코드 보강
